certbot-ci/certbot_integration_tests/utils/dns_server.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python
+"""Module to setup an RFC2136-capable DNS server"""
+import os
+import os.path
+from pkg_resources import resource_filename
+import shutil
+import socket
+import subprocess
+import sys
+import tempfile
+import time
+
+
+BIND_DOCKER_IMAGE = 'internetsystemsconsortium/bind9:9.16'
+BIND_BIND_ADDRESS = ('127.0.0.1', 45953)
+
+# A TCP DNS message which is a query for '. CH A' transaction ID 0xcb37. This is used
+# by _wait_until_ready to check that BIND is responding without depending on dnspython.
+BIND_TEST_QUERY = bytearray.fromhex('0011cb37000000010000000000000000010003')
+
+
+class DNSServer(object):
+    """
+    DNSServer configures and handles the lifetime of an RFC2136-capable server.
+    DNServer provides access to the dns_xdist parameter, listing the address and port
+    to use for each pytest node.
+
+    At this time, DNSServer should only be used with a single node, but may be expanded in
+    future to support parallelization (https://github.com/certbot/certbot/issues/8455).
+    """
+
+    def __init__(self, nodes, show_output=False):
+        """
+        Create an DNSServer instance.
+        :param list nodes: list of node names that will be setup by pytest xdist
+        :param bool show_output: if True, print the output of the DNS server
+        """
+
+        self.bind_root = tempfile.mkdtemp()
+
+        self.dns_xdist = {
+            'address': BIND_BIND_ADDRESS[0],
+            'port': BIND_BIND_ADDRESS[1]
+        }
+
+        # Unfortunately the BIND9 image forces everything to stderr with -g and we can't
+        # modify the verbosity.
+        self._output = sys.stderr if show_output else open(os.devnull, 'w')
+
+    def start(self):
+        """Start the DNS server"""
+        try:
+            self._configure_bind()
+            self._start_bind()
+        except:
+            self.stop()
+            raise
+
+    def stop(self):
+        """Stop the DNS server, and clean its resources"""
+        try:
+            self.process.terminate()
+            self.process.wait()
+        except BaseException as e:
+            print("BIND9 did not stop cleanly: {}".format(e), file=sys.stderr)
+            pass
+        finally:
+            shutil.rmtree(self.bind_root)
```

Comment:
This call should also be protected by `try/except` to avoid to polute the logs when the temporary workspace has not even been setup before the main exception in the process occurs.